python/cudf_polars/cudf_polars/experimental/base.py

```diff
@@ -248,16 +250,61 @@ def new_parent(
         )
 
 
+class JoinKey:
+    """
+    Join-key information.
+
+    Parameters
+    ----------
+    column_stats
+        Column statistics for the join key.
+
+    Notes
+    -----
+    This class is used to track join-key information.
+    It is used to track the columns being joined on
+    and the estimated unique-value count for the join key.
+    """
+
+    column_stats: tuple[ColumnStats, ...]
+    unique_count_estimate: int | None
+    """Estimated unique-value count from join heuristics."""
+
+    def __init__(self, *column_stats: ColumnStats) -> None:
+        self.column_stats = column_stats
+        self.unique_count_estimate = None
+
+    @cached_property
+    def source_row_count(self) -> int | None:
+        """Return the estimated row-count of the source columns."""
+        return min(
+            (
+                cs.source_info.row_count.value
+                for cs in self.column_stats
+                if cs.source_info.row_count.value is not None
+            ),
+            default=None,
+        )
+
+
 class StatsCollector:
     """Column statistics collector."""
 
-    __slots__ = ("column_stats", "row_count")
+    __slots__ = ("column_stats", "join_keys", "joins", "row_count")
 
     row_count: dict[IR, ColumnStat[int]]
     """Estimated row count for each IR node."""
     column_stats: dict[IR, dict[str, ColumnStats]]
     """Column statistics for each IR node."""
+    join_keys: MutableMapping[JoinKey, set[JoinKey]]
+    """Join-key mappings."""
```

Comment:
Docs / a more descriptive name on what the relationship between the key and values here would be helpful. It's not immediately obvious to me why we're mapping `JoinKeys` to (sets of) `JoinKeys`.